design/ceph-osd-redesign-for-octopus.md

```diff
@@ -0,0 +1,342 @@
+
+This option is similar to Rook's current usage of directory-based OSDs, but it is critically
+different in that it is explicitly reserved for test environment usage and is not made to be as
+configurable or flexible as the supported disk usage. This inflexibility is important for keeping
+the new OSD implementation as clean, extensible, and maintainable as possible for production
+features.
+
+#### PVCs as backing storage
+The new code path must support using PVCs as backing storage, which is a design that is currently in
+progress. PVCs may (or may not) be an an exception to the previous mental model changes. Rook must
+be able to query nodes for storage devices, but it may not be able to query PVCs which may be
+claimed for Rook's use.
+
+#### Operator needs knowledge of OSD host hardware
+Currently, the Rook operator has limited knowledge of the hardware being configured on hosts.
+Hardware knowledge is confined to the OSD preparation binary. Mechanisms are not currently in place
+to report the needed disk information to the operator. The updated config plan calls for a Ceph
+config file that is minimal and standard for all daemons; daemons are individually configured using
+command line flags. In order to properly set command line flags for object storage daemons, the
+operator must have knowledge of the disks which an OSD will use. The operator must be able to
+configure, for example, an OSD to use one disk as the main storage, another disk as wal storage, and
+another disk as db storage. The new design will give the operator requisite knowledge.
+
+
+### Proposed Rook operator design
+The existing OSD operator design uses a two-stage process. The first stage is preparation of OSDs on
+a per-host basis, and the second stage is running individual OSDs. The design proposed here
+separates OSD tasks into more stages.
+
+A Rook OSD prepare binary currently handles the equivalent of stages 1 through 4, and in early
+implementations of the new OSD design, a version of the current binary could be used for stages 1
+through 3 to achieve the minimum viable product more quickly. Splitting the stages can be done
+afterwards as makes sense.
+
+#### Stage 1: Host device inventory
+The first stage is an inspection stage that is currently done as part of the existing preparation
+stage. Similar to the current Rook OSD prepare binary, this proposal calls for a Rook inspection
+binary which will examine the disk hardware, rejecting disks that are already in use by applications
+other than Rook (as per current behavior). Rook currently uses `lsblk` and custom logic to find free
+disks, but this design proposes to use `ceph-volume inventory` to find free disks. The inspection
+binary stores the result of the `ceph-volume inventory` command to a ConfigMap which  will be read
+by the operator.
+
+#### Stage 2: Process device inspection report
+In the second stage, the operator examines the host device inspection report in the ConfigMap. The
```

Comment:
As part of that, I think we should also analyze for a given device that it is either `available` or `used` by the current cluster fsid.

Reply:
I'm not quite sure I understand the distinction you're making here.

Reply:
I guess what I was trying to say is that if someone declares `/dev/sda` and we find `/dev/sda` is not available but by looking at the reason we figured out it's because it's already an OSD own by our cluster then it's good. Hope it makes sense.

Reply:
Yes. That makes sense now.